lib/checksizeof.cpp

```diff
@@ -373,16 +373,20 @@ void CheckSizeof::suspiciousSizeofCalculation()
     if (!mSettings->severity.isEnabled(Severity::warning) || !mSettings->certainty.isEnabled(Certainty::inconclusive))
         return;
 
-    // TODO: Use AST here. This should be possible as soon as sizeof without brackets is correctly parsed
     for (const Token *tok = mTokenizer->tokens(); tok; tok = tok->next()) {
         if (Token::simpleMatch(tok, "sizeof (")) {
-            const Token* const end = tok->linkAt(1);
-            const Variable* var = end->previous()->variable();
-            if (end->strAt(-1) == "*" || (var && var->isPointer() && !var->isArray())) {
-                if (end->strAt(1) == "/")
-                    divideSizeofError(tok);
-            } else if (Token::simpleMatch(end, ") * sizeof") && end->next()->astOperand1() == tok->next())
-                multiplySizeofError(tok);
+            const Token* lPar = tok->astParent();
+            if (lPar && lPar->str() == "(") {
+                const Token* const end = lPar->link();
+                const Token* const varTok = end->previous();
+                const Variable* var = varTok ? varTok->variable() : nullptr;
+                if ((varTok && varTok->str() == "*") || (var && var->isPointer() && !var->isArray() && !(lPar->astOperand2() && lPar->astOperand2()->str() == "*"))) {
```

Comment:
this is better but not perfect. It will not deal with all expressions properly. I'd assume `sizeof(ptr+1)` should be treated as a ptr expression for instance. I feel it can be merged like this.. but ideally somebody could look into it some more. if you don't want to look at it in the near future.. how about a TODO comment.

Reply:
I had thought about using `nextAfterAstRightmostLeaf()` instead of `end->previous()`. Maybe I can come up with a test case.